src/backend/distributed/planner/merge_planner.c

```diff
@@ -736,3 +732,31 @@ MergeQualAndTargetListFunctionsSupported(Oid resultRelationId, FromExpr *joinTre
 
 
 #endif
+
+
+/*
+ * IsLocalTableModification returns true if the table modified is a Postgres table.
+ */
+bool
+IsLocalTableModification(Query *query, uint64 shardId, RTEListProperties *rteProperties)
+{
+	/* No-op for SELECT command */
+	if (!IsModifyCommand(query))
+	{
+		return false;
+	}
+
+	/* For MERGE, we have to check only the target relation */
+	if (IsMergeQuery(query) && !IsCitusTable(targetRelationId))
```

Comment:
Why merge is different from, say UPDATE with subquery for this check?

Reply:
What happened is, in MERGE, if we have Postgres table as **source** tables, it's incorrectly flagging as localtable modification, though modifying Citus table. Later, in the execution phase, this flag is used to generate the tasklist based on shards, though the target is Citus shard, it's skipping generating task list in  **CitusBeginModifyScan**

Reply:
> it's skipping generating task list in CitusBeginModifyScan,

As far as I can see, this function never generates the task list, the task list is always there. Can you please clarify the meaning of "generating task list"


> we have Postgres table as source tables, it's incorrectly flagging as localtable modification,

Hmm, even if I undo the code:
```C
diff --git a/src/backend/distributed/planner/merge_planner.c b/src/backend/distributed/planner/merge_planner.c
index 90b4e530d..0273a8504 100644
--- a/src/backend/distributed/planner/merge_planner.c
+++ b/src/backend/distributed/planner/merge_planner.c
@@ -727,7 +727,7 @@ IsLocalTableModification(Oid targetRelationId, Query *query, uint64 shardId,
        }

        /* For MERGE, we have to check only the target relation */
-       if (IsMergeQuery(query) && !IsCitusTable(targetRelationId))
+       if (false && IsMergeQuery(query) && !IsCitusTable(targetRelationId))
        {
                /* Postgres table */
                return true;
```

I can see that this works, where `reftarget_local ` is Citus local, and `reftarget_pglocal ` is pg local. Can you maybe share the steps that cause the problem you mentioned?

```SQL
 explain MERGE INTO reftarget_local tar
USING (SELECT * FROM reftarget_pglocal) AS foo ON tar.t1 = foo.t1
WHEN MATCHED AND tar.t2 = 100 THEN
        DELETE
WHEN MATCHED THEN
        UPDATE SET t2 = foo.t2 + 100
WHEN NOT MATCHED THEN
        INSERT VALUES(foo.t2);
```




> though the target is Citus shard

Is it possible that we do not set the task->anchorShardId field properly, which causes the problem?


> Why merge is different from, say UPDATE with subquery for this check?

I'm double clicking this because I think the same problem should exists for other queries such as `UPDATE citus_local FROM pg_local` etc. They seem to go through the same code-path. I'd like to make sure we do not skip any wrong behavior on such cases, but I cannot see the problem either.

Reply:
Here is the unit test

```
DROP SCHEMA test cascade;
SET citus.shard_count TO 4;
create schema test;
set search_path to 'test';

CREATE TABLE pg(t1 int, t2 int);
CREATE TABLE reference(s1 int, s2 int);

INSERT INTO pg VALUES(1, 0);
INSERT INTO pg VALUES(3, 100);
INSERT INTO reference VALUES(1, 1);
INSERT INTO reference VALUES(2, 2);
INSERT INTO reference VALUES(3, 3);

SELECT create_reference_table('reference');

set citus.log_remote_commands to true;
MERGE INTO pg
USING (SELECT * FROM reference UNION SELECT * FROM reference) AS foo ON pg.t1 = foo.s1
WHEN MATCHED AND pg.t2 = 100 THEN
        DELETE
WHEN MATCHED THEN
        UPDATE SET t2 = t2 + 100
WHEN NOT MATCHED THEN
        INSERT VALUES(foo.s1);

```
I removed my fix

```
#if 0
        if (IsMergeQuery(query) && !IsCitusTable(targetRelationId))
        {
                /* Postgres table */
                return true;
        }
#endif
```

You can see the failure

```
psql:refer-source:26: NOTICE:  issuing SELECT lock_shard_resources(7, ARRAY[102018])
DETAIL:  on server citus@127.0.0.1:10000 connectionId: 2
psql:refer-source:26: NOTICE:  issuing MERGE INTO test.pg USING (SELECT reference.s1, reference.s2 FROM test.reference_102018 reference UNION SELECT reference.s1, reference.s2 FROM test.reference_102018 reference) foo ON (pg.t1 OPERATOR(pg_catalog.=) foo.s1) WHEN MATCHED AND (pg.t2 OPERATOR(pg_catalog.=) 100) THEN DELETE WHEN MATCHED THEN UPDATE SET t2 = (pg.t2 OPERATOR(pg_catalog.+) 100) WHEN NOT MATCHED THEN INSERT (t1) VALUES (foo.s1)
DETAIL:  on server citus@127.0.0.1:10000 connectionId: 2
psql:refer-source:26: NOTICE:  issuing BEGIN TRANSACTION ISOLATION LEVEL READ COMMITTED;SELECT assign_distributed_transaction_id(0, 4, '2023-04-25 11:34:44.419385-07');
DETAIL:  on server citus@127.0.0.1:11000 connectionId: 1
psql:refer-source:26: ERROR:  relation "test.pg" does not exist
CONTEXT:  while executing command on 127.0.0.1:10000
psql:refer-source:26: NOTICE:  issuing ROLLBACK
DETAIL:  on server citus@127.0.0.1:10000 connectionId: 2

```

With the fix

```
psql:refer-source:26: NOTICE:  issuing SELECT lock_shard_resources(7, ARRAY[102019])
DETAIL:  on server citus@127.0.0.1:10000 connectionId: 2
psql:refer-source:26: NOTICE:  executing the command locally: MERGE INTO test.pg USING (SELECT reference.s1, reference.s2 FROM test.reference_102019 reference UNION SELECT reference.s1, reference.s2 FROM test.reference_102019 reference) foo ON (pg.t1 OPERATOR(pg_catalog.=) foo.s1) WHEN MATCHED AND (pg.t2 OPERATOR(pg_catalog.=) 100) THEN DELETE WHEN MATCHED THEN UPDATE SET t2 = (pg.t2 OPERATOR(pg_catalog.+) 100) WHEN NOT MATCHED THEN INSERT (t1) VALUES (foo.s1)
psql:refer-source:26: NOTICE:  issuing PREPARE TRANSACTION 'citus_0_1457226_4_7'
DETAIL:  on server citus@127.0.0.1:10000 connectionId: 2
psql:refer-source:26: NOTICE:  issuing COMMIT PREPARED 'citus_0_1457226_4_7'
DETAIL:  on server citus@127.0.0.1:10000 connectionId: 2
MERGE 3
```

Reply:
@onderkalaci There is a mistake in my earlier comment, the modification table i.e. target should be PG and not source.

Reply:
BTW, that check for `shardId == INVALID_SHARD_ID` is confusing, I believe, it was there to eliminate **citus-local** tables, as the second condition `ContainsOnlyLocalTables` is only checking for distributed and reference tables.

My best guess is that, originally code has only `ContainsOnlyLocalTables` and as we introduced citus-local tables, the author might have appended `shardId == INVALID_SHARD_ID`?

Reply:
> Why merge is different from, say UPDATE with subquery for this check?

Alright, now I understand the difference. Thanks for the explanation.

I think the main difference is that for `UPDATE/DELETE` we follow recursive planning, which eliminates any joins between local tables and distributed/reference tables:
```SQL
update pg SET t1 = reference.s1 FROM reference WHERE reference.s2 = pg.t2 ;
NOTICE:  executing the command locally: SELECT s1, s2 FROM test.reference_102008 reference WHERE true
NOTICE:  executing the command locally: UPDATE test.pg SET t1 = reference.s1 FROM (SELECT reference_1.s1, reference_1.s2 FROM (SELECT intermediate_result.s1, intermediate_result.s2 FROM read_intermediate_result('11_1'::text, 'binary'::citus_copy_format) intermediate_result(s1 integer, s2 integer)) reference_1) reference WHERE (reference.s2 OPERATOR(pg_catalog.=) pg.t2)
```


For `Merge` we do not support recursive planning at the moment. So, would be nice to have a comment on that here, what makes the difference.

The recursive logic kicks in here https://github.com/citusdata/citus/blob/efd41e8ea55cf613fc1a0255a034ce13ceb554d3/src/backend/distributed/planner/local_distributed_join_planner.c#L195-L219

When we support `subqueries/ctes`, I think we'll get through this logic automatically. 

For now, it seems fine to skip that after adding